frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -342,7 +342,7 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
     Priority requestPriority = statusManager.getNextContainerRequestPriority();
     String requestNodeLabel = requestManager.getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
 
-    ResourceDescriptor requestResource = requestManager.getTaskResources().get(taskRoleName);
+    ResourceDescriptor requestResource = ResourceDescriptor.formatResource(requestManager.getTaskResources().get(taskRoleName));
```

Comment:
>ResourceDescriptor.formatResource [](start = 41, length = 33)

Do not format again and again here.
Move the call into RequestManager.pullRequest, like flattenFrameworkDescriptor.
You can add function like reviseFrameworkDescriptor.

Reply:
Only format resource once when requests received from web server?

---
In reply to: [186061412](https://github.com/Microsoft/pai/pull/559#discussion_r186061412) [](ancestors = 186061412)

Reply:
It is "Only format resource one time when new requests pulled from Zookeeper."

---
In reply to: [186331808](https://github.com/Microsoft/pai/pull/559#discussion_r186331808) [](ancestors = 186331808,186061412)